### PR TITLE
Fermat

### PR DESCRIPTION
**Pull request by jhlchan**
Update algebra library, incorporate an improved algorithm for Fermat's Two Squares Theorem.

### CODE REVIEW

File: examples/fermat/twosq/cornerScript.sml [hunk 1]
```diff
@@ -0,0 +1,1075 @@
...
+(* Proof:
+   Let s = corners_less_fix p.
+   Note s <> {}                                by tik_prime_corners_less_fix
+    ==> ?q. q IN s                             by MEMBER_NOT_EMPTY
+    ==> ?a b. q = (a,a,b,b) /\
+              p = a ** 2 + b ** 2 /\ a < b     by corners_less_fix_def
+   Take them and form a pair (a,b).
+*)
+Theorem fermat_two_squares_by_corners:
+  !p. tik p /\ prime p ==> ?(a, b). p = a ** 2 + b ** 2 /\ a < b
```
Comment (mn200) [inline examples/fermat/twosq/cornerScript.sml:1030]: Unless there's strong reason to, please don't use paired existentials, like `?(a,b)....` Instead, just have the existentially quantified variables listed: `?a b. ...`

Comment (jhlchan) [reply examples/fermat/twosq/cornerScript.sml]: New commit uses simple existentially quantified variables. Should I worry about the failure of [d64379b]?

**Comment by mn200** [comment]
Failure (visible through link) is:

```
#13 3771.2  ∀n. n MOD 4 ≠ 0 ⇒ fixes zagier (mills n) = {(x,x,z) | n = windmill (x,x,z)}
#13 3771.2  
#13 3771.2  failed.
#13 3771.2  Failed to prove theorem zagier_fixes.
#13 3771.2  
```

This was in the experimental kernel where variable names chosen by the kernel can be different.

File: examples/fermat/twosq/cornerScript.sml (hunk 1 above)
Comment (jhlchan) [reply examples/fermat/twosq/cornerScript.sml]: Got it. Now I learn how to read the log, and fixed the problem. Hope this is the only one.

**Comment by mn200** [comment]
Still seem to be an issue?

File: examples/fermat/twosq/cornerScript.sml (hunk 1 above)
Comment (jhlchan) [reply examples/fermat/twosq/cornerScript.sml]: This time it is the standard kernel, but I have used rename1 to ensure proper variable rename. Log seems not to fail in fermat/ folder. Any suggestion about the exact issue?

**Comment by mn200** [comment]
That more recent error is not yours; it's a common repeated error that that file throws up non-deterministically.

Thanks for your work!

File: examples/fermat/twosq/cornerScript.sml (hunk 1 above)
Comment (jhlchan) [reply examples/fermat/twosq/cornerScript.sml]: Thanks for all your help, much appreciated.

**Pull request merged**